Improve internal DX around byte classification [1]

RobinMalfait opened the pull request:
This PR improves the internal DX when working with `u8` classification into a smaller enum. This is done by implementing a `ClassifyBytes` proc derive macro. The benefit of this is that the DX is much better and everything you will see here is done at compile time.

Before:
```rs
#[derive(Debug, Clone, Copy, PartialEq)]
enum Class {
    ValidStart,
    ValidInside,
    OpenBracket,
    OpenParen,
    Slash,
    Other,
}

const CLASS_TABLE: [Class; 256] = {
    let mut table = [Class::Other; 256];

    macro_rules! set {
        ($class:expr, $($byte:expr),+ $(,)?) => {
            $(table[$byte as usize] = $class;)+
        };
    }

    macro_rules! set_range {
        ($class:expr, $start:literal ..= $end:literal) => {
            let mut i = $start;
            while i <= $end {
                table[i as usize] = $class;
                i += 1;
            }
        };
    }

    set_range!(Class::ValidStart, b'a'..=b'z');
    set_range!(Class::ValidStart, b'A'..=b'Z');
    set_range!(Class::ValidStart, b'0'..=b'9');

    set!(Class::OpenBracket, b'[');
    set!(Class::OpenParen, b'(');

    set!(Class::Slash, b'/');

    set!(Class::ValidInside, b'-', b'_', b'.');

    table
};
```

After:
```rs
#[derive(Debug, Clone, Copy, PartialEq, ClassifyBytes)]
enum Class {
    #[bytes_range(b'a'..=b'z', b'A'..=b'Z', b'0'..=b'9')]
    ValidStart,

    #[bytes(b'-', b'_', b'.')]
    ValidInside,

    #[bytes(b'[')]
    OpenBracket,

    #[bytes(b'(')]
    OpenParen,

    #[bytes(b'/')]
    Slash,

    #[fallback]
    Other,
}
```

Before we were generating a `CLASS_TABLE` that we could access directly, but now it will be part of the `Class`. This means that the usage has to change:

```diff
- CLASS_TABLE[cursor.curr as usize]
+ Class::TABLE[cursor.curr as usize]
```

This is slightly worse UX, and this is where another change comes in. We implemented the `From<u8> for #enum_name` trait inside of the `ClassifyBytes` derive macro. This allows us to use `.into()` on any `u8` as long as we are comparing it to a `Class` instance. In our scenario:

```diff
- Class::TABLE[cursor.curr as usize]
+ cursor.curr.into()
```

Usage wise, this looks something like this:
```diff
        while cursor.pos < len {
-           match Class::TABLE[cursor.curr as usize] {
+           match cursor.curr.into() {
-               Class::Escape => match Class::Table[cursor.next as usize] {
+               Class::Escape => match cursor.next.into() {
                    // An escaped whitespace character is not allowed
                    Class::Whitespace => return MachineState::Idle,

                    // An escaped character, skip ahead to the next character
                    _ => cursor.advance(),
                },

                // End of the string
                Class::Quote if cursor.curr == end_char => return self.done(start_pos, cursor),

                // Any kind of whitespace is not allowed
                Class::Whitespace => return MachineState::Idle,

                // Everything else is valid
                _ => {}
            };

            cursor.advance()
        }

        MachineState::Idle
    }
}
```


If you manually look at the `Class::TABLE` in your editor for example, you can see that it is properly generated at compile time.

Given this input:
```rs
#[derive(Clone, Copy, ClassifyBytes)]
enum Class {
    #[bytes_range(b'a'..=b'z')]
    AlphaLower,

    #[bytes_range(b'A'..=b'Z')]
    AlphaUpper,

    #[bytes(b'@')]
    At,

    #[bytes(b':')]
    Colon,

    #[bytes(b'-')]
    Dash,

    #[bytes(b'.')]
    Dot,

    #[bytes(b'\0')]
    End,

    #[bytes(b'!')]
    Exclamation,

    #[bytes_range(b'0'..=b'9')]
    Number,

    #[bytes(b'[')]
    OpenBracket,

    #[bytes(b']')]
    CloseBracket,

    #[bytes(b'(')]
    OpenParen,

    #[bytes(b'%')]
    Percent,

    #[bytes(b'"', b'\'', b'`')]
    Quote,

    #[bytes(b'/')]
    Slash,

    #[bytes(b'_')]
    Underscore,

    #[bytes(b' ', b'\t', b'\n', b'\r', b'\x0C')]
    Whitespace,

    #[fallback]
    Other,
}
```

This is the result:
<img width="1244" alt="image" src="https://github.com/user-attachments/assets/6ffd6ad3-0b2f-4381-a24c-593e4c72080e" />

RobinMalfait committed:
add `classification-macro`

This is a little proc macro that allows us to write the `enum Class` in
a more ergonomic way.

RobinMalfait committed:
add `u8` extension to the `ClassifyBytes` enum

This allows us to apply this change:

```diff
- Class::CLASS_TABLE[cursor.curr as usize]
+ cursor.curr.classify()
```

Where `cursor.curr` is a `u8`

RobinMalfait committed:
use new `ClassifyBytes` proc derive macro in extractor machines

RobinMalfait committed:
use `u8.classify()` method

The `classify` method is implemented as `Class::CLASS_TABLE[*self as
usize]` and uses an `#[inline(always)]` as well.

This is just a DX improvement.

RobinMalfait committed:
use shorter `TABLE` name

This gets rid of the redundant `Class::CLASS_TABLE`

```diff
- Class::CLASS_TABLE
+ Class::TABLE
```

RobinMalfait committed:
implement `From<u8> for #enum_name` instead of `classify()`

There is a tiny problem with the custom trait which exposes
`classify()`. The moment you have 2 enums with `ClassifyBytes` in the
same file, it means that the `trait U8Ext` was duplicated resulting in
errors.

To solve this, we can also implement the `From<u8> for #enum_name`
trait. This allows us to just use `b'a'.into()` in a context where you
are comparing it with an `Class::AlphaLower` for example.

RobinMalfait committed:
organize imports

thecrypticace commented on crates/classification-macros/src/lib.rs:
```diff
@@ -103,15 +103,9 @@ pub fn classify_bytes_derive(input: TokenStream) -> TokenStream {
-            fn classify(&self) -> #enum_name {
-                #enum_name::TABLE[*self as usize]
+        impl From<u8> for #enum_name {
+            fn from(byte: u8) -> Self {
```
I guess this will always be inlined by Rust just b/c small (Rust v1.75+ change)?

RobinMalfait replied:
Didn't check but also didn't notice any perf impact at all.

RobinMalfait committed:
use correct machine in performance test

RobinMalfait committed:
declaration properties can only contain dashes and a-z values

In case the property starts with `--`, we will parse it as a CSS
variable which can contain a few more characters like emoji for example.

Initial reason that I included uppercase is because I was thinking about
`Webkit`, but that should be `-webkit` in CSS anyway.

Technically uppercase values are valid in CSS:
```css
BODY {
  BACKGROUND-COLOR: RED;
}
```
Play: https://play.tailwindcss.com/sUsWi0Ampe?file=css

But for the sake of the arbitrary property, let's only allow
lower-kebab-case property names.

philipp-spiess commented on crates/classification-macros/Cargo.toml:
https://doc.rust-lang.org/reference/procedural-macros.html#r-macro.proc.def ah. should have googled first huh 😆

RobinMalfait replied:
Hehe yep 👍

RobinMalfait committed:
Merge branch 'main' into feat/improve-classification

philipp-spiess reviewed (approved):
Awesome stuff. Happy to see the CLASS_TABLE macro stuff gone honestly. Not really happy that this has to be its own crate but what can you do. Also unsure wether `Char#into()` is super intuitive API but kinda also nice that it doesn't leak that it's a lookup table internally so maybe that's cool. Anyway no blockers from my side 👍

RobinMalfait commented on crates/oxide/src/extractor/arbitrary_property_machine.rs:
```diff
@@ -68,7 +69,7 @@ impl Machine for ArbitraryPropertyMachine {
 
         match self.state {
-            State::Idle => match CLASS_TABLE[cursor.curr as usize] {
+            State::Idle => match cursor.curr.into() {
```
Throughput wise this is indeed the same because everything is inlined. What I've seen so far is that in Rust people rely a lot on these traits and converting from a `u8` into a `Class` is only possible if the trait is imported so at least there is no magic where it's coming from.

Implementing the `From` trait to get the `into()` method is something I see a lot in the wild. Similar to the `Default` trait and I personally like it 🤔 

In this particular case, since it's coming via a macro, if you go to definition on the `into()` call you will see it jump to the `Class` enum (or if you get the option in your editor you can also jump to the macro itself)

If you use `hover` on the `into()` you also see the `Class` type:
<img width="808" alt="image" src="https://github.com/user-attachments/assets/f7e87438-53fc-446f-b01e-789e14a38135" />

RobinMalfait commented on crates/oxide/src/extractor/arbitrary_property_machine.rs:
```diff
@@ -102,7 +103,7 @@ impl Machine for ArbitraryPropertyMachine {
                         // E.g.: `[color:red]`
                         //         ^^^^^
-                        Class::Alpha => cursor.advance(),
+                        Class::AlphaLower => cursor.advance(),
```
🤫

RobinMalfait commented on crates/classification-macros/src/lib.rs:
```diff
@@ -0,0 +1,248 @@
+        }
+    }
+
+    // If no fallback variant is found, default to "Other"
+    let fallback_ident = fallback_variant.unwrap_or_else(|| Ident::new("Other", Span::call_site()));
```
Yeah we can be explicit here; In this PR I've always added the `#[fallback]` anyway. Let me change it.

RobinMalfait commented on crates/classification-macros/src/lib.rs:
```diff
@@ -0,0 +1,248 @@
+        if attr.path().is_ident("bytes") {
+            match parse_bytes_attr(attr) {
+                Ok(list) => assigned.extend(list),
+                Err(e) => panic!("Error parsing #[bytes(...)]: {}", e),
```
The `parse_bytes_attr` is using a `Result` coming from the `syn` crate already.

So right now if you do something incorrect, you will see an error like this:
```
  --> crates/oxide/src/extractor/string_machine.rs:71:45
   |
71 | #[derive(Debug, Clone, Copy, PartialEq, Eq, ClassifyBytes)]
   |                                             ^^^^^^^^^^^^^
   |
   = help: message: Error parsing #[bytes(...)]: Expected a byte literal like b'a'
```

We can use the default `Result` type, which requires a change like this:
```diff
- Result<Vec<u8>>
- Result<Vec<u8>, syn::Error>
```

But it results in the same kind of error:
```
  --> crates/oxide/src/extractor/string_machine.rs:71:45
   |
71 | #[derive(Debug, Clone, Copy, PartialEq, Eq, ClassifyBytes)]
   |                                             ^^^^^^^^^^^^^
   |
   = help: message: Error parsing #[bytes(...)]: Expected a byte literal like b'a'

```

So not sure if it's worth it, especially because `syn` already has a `Result` enum shorthand.

RobinMalfait committed:
hard panic when no `#[fallback]` is found

If you're missing the `#[fallback]` variant, you will see an error like
this:

```
error: proc-macro derive panicked
  --> crates/oxide/src/extractor/string_machine.rs:71:45
   |
71 | #[derive(Debug, Clone, Copy, PartialEq, Eq, ClassifyBytes)]
   |                                             ^^^^^^^^^^^^^
   |
   = help: message: A variant marked with #[fallback] is missing
```
